contrib/mod_tls.c

```diff
@@ -13005,6 +13002,81 @@ MODRET tls_any(cmd_rec *cmd) {
   return PR_DECLINED(cmd);
 }
 
+/* Issue #1931: For TLSv1.3 sessions, we should use SSL_new_session_ticket
+ * to request a new ticket (and thus session) before any data transfers.
+ *
+ * For TLSv1.2 sessions with (and without) tickets, we might try
+ * SSL_renegotiate in the future.
+ *
+ * TODO: Work on solutions for TLSv1.2 and other protocol versions.
+ */
+MODRET tls_pre_xfer(cmd_rec *cmd) {
+#if defined(TLS1_3_VERSION)
+  SSL_SESSION *ssl_session;
+  unsigned long sess_remaining;
+  int request_new_ticket = FALSE, ssl_version;
+
+  if (tls_engine == FALSE) {
+    return PR_DECLINED(cmd);
+  }
+
+  if (!(tls_flags & TLS_SESS_ON_CTRL)) {
+    return PR_DECLINED(cmd);
+  }
+
+  if (ctrl_ssl == NULL) {
+    return PR_DECLINED(cmd);
+  }
+
+  ssl_session = SSL_get_session(ctrl_ssl);
+  ssl_version = SSL_SESSION_get_protocol_version(ssl_session);
+
+  if (ssl_version != TLS1_3_VERSION) {
+    return PR_DECLINED(cmd);
+  }
+
+  sess_remaining = tls_sess_remaining(ssl_session);
+  pr_trace_msg(trace_channel, 21,
+    "%s: control channel %s session lifetime: %lu %s%s",
+    (const char *) cmd->argv[0], SSL_get_version(ctrl_ssl), sess_remaining,
+    sess_remaining != 1 ? "seconds" : "second",
+    sess_remaining != 0 ? "" : " (EXPIRED)");
+
+  /* If the control session expires in 10 seconds or less, request a new one
+   * before proceeding with the data transfer (Issue #1931).
+   */
+  if (sess_remaining <= 10) {
+    request_new_ticket = TRUE;
+  }
+
+  if (request_new_ticket == TRUE) {
+    int res;
+
+# if defined(OPENSSL_VERSION_MAJOR) && \
+     OPENSSL_VERSION_MAJOR >= 3
+    res = SSL_new_session_ticket(ctrl_ssl);
```

Comment:
I recently discovered that there is another solution possible, when using OpenSSL-1.x _or_ LibreSSL: we can extend the lifetime of the current TLS session object, such that it will be valid (not expired) when the subsequent data transfer TLS handshake occurs, such that `SSL_session_reused` will return true.

So this PR is not quite done yet.

However, that still leaves some edge cases to be explored: OpenSSL-3.x and non-TLSv1.3 sessions, with and without session tickets.